depends/toolchain.cmake.in

```diff
@@ -4,8 +4,13 @@
 
 # This file is expected to be highly volatile and may still change substantially.
 
-set(CMAKE_SYSTEM_NAME @host_system@)
-set(CMAKE_SYSTEM_PROCESSOR @host_arch@)
+set(BITCOIN_BUILD_WITH_DEPENDS TRUE)
```

Comment:
I'm not sure a `BITCOIN_BUILD_WITH_DEPENDS` carve out is the way to go (unless this is going to be removed?); as that would imply that we now have essentially 3 different build modes? "Normal" compilation, depends (which produces a toolchain file), and then any other external toolchain file, and things may be slightly different between all 3?

If our goal is to ultimately align everything on CMake toolchain files, and depends is going to produce a CMake toolchain file, then we should treat it the same as any other, not have a special carve out in a build system, where certain toolchain files are treated differently to others.

Reply:
I'll add, I'm not exactly sure what the right solution is, but my point in https://github.com/hebasto/bitcoin/pull/82#discussion_r1534141067, was that it was confusing to have code branching based on `CMAKE_CROSSCOMPILING`, when that didn't actually always mean cross-compilation (I guess there was a bug there too), and I'm not sure that this change makes things a lot clearer, because it now makes depends "special", when it really shouldn't be.

Reply:
> I'm not sure a `BITCOIN_BUILD_WITH_DEPENDS` carve out is the way to go (unless this is going to be removed?)

Yes, it might be removed. Probably, the most problematic case is related to finding the `qt` package. At this point, this variable simplifies conveying the depends-specific variables to the main build system.

> that would imply that we now have essentially 3 different build modes? "Normal" compilation, depends (which produces a toolchain file), and then any other external toolchain file, and things may be slightly different between all 3?

At the highest level, I can see two major conditions, each of which has two distinct states:
1. A toolchain file is provided or not.
2. A cross-compiling is triggered or not.

So, there are 4 different build modes. For example, cross-compiling might be triggered by a user even without a toolchain file.

But this way to estimate the number of the possible build system states / modes does not seem productive, as there are more of them, for example with / without wallet.

> If our goal is to ultimately align everything on CMake toolchain files, and depends is going to produce a CMake toolchain file, then we should treat it the same as any other, not have a special carve out in a build system, where certain toolchain files are treated differently to others.

I don't agree with this point. If a toolchain file sets a variable, which affects the main build system behaviour, then it does not mean that this toolchain file is treated differently. Basically, it's toolchain file's responsibility to set variables that alters the main build system behavior.


> my point in [#82 (comment)](https://github.com/hebasto/bitcoin/pull/82#discussion_r1534141067), was that it was confusing to have code branching based on `CMAKE_CROSSCOMPILING`, when that didn't actually always mean cross-compilation (I guess there was a bug there too)...

I agree. And this PR resolves this issue.

> I'm not sure that this change makes things a lot clearer, because it now makes depends "special", when it really shouldn't be.

Depends are "special" only in a way that they define some custom variables. For example, Android toolchain files provide a bunch of Android-specific variables as well. I can't see any drawback here.

Reply:
Reworked.